src/parse/rules/go_rules.build_defs

```diff
@@ -512,114 +514,140 @@ def cgo_test(name:str, srcs:list, data:list=None, deps:list=None, visibility:lis
+      hashes (list): List of hashes to verify the downloaded sources against.
     """
-    post_build = None
-    if binary and outs and len(outs) != 1:
-        raise ValueError(name + ': Binary rules can only have a single output')
+    if hashes and not revision:
+        log.warning("You shouldn't specify hashes on go_get without specifying revision as well")
+    labels = ['go_get:%s@%s' % (get, revision) if revision else 'go_get:%s' % get]
     getroot = get[:-4] if get.endswith('/...') else get
     subdir = 'src/' + getroot
-    if not outs:
-        outs = [('bin/' + name) if binary else ('src/' + getroot)]
-        if not binary:
-            post_build = _extra_outs(getroot)
+
+    # Some special optimisation for github, which lets us download zipfiles at a particular sha instead of
+    # cloning the whole repo. Obviously that is a lot faster than cloning entire repos.
+    if get.startswith('github.com') and revision and get.count('/') >= 2:
+        cmd = _go_github_repo_cmd(getroot, getroot, revision)
+    elif get.startswith('golang.org/x/') and revision and not repo:
+        # We know about these guys...
+        cmd = _go_github_repo_cmd(getroot, 'github.com/golang/' + getroot[len('golang.org/x/'):], revision)
+    else:
+        if repo:
+            # we've been told exactly where to get the source from.
+            cmd = ['git clone %s %s' % (repo, getroot)]
+        else:
+            # Ultimate fallback to go get.
+            # This has some more restrictions than the above (e.g. go get won't fetch a directory
+            # with no Go files in it, even if passed -d).
+            cmd = [
+                'rm -rf src pkg',
+                'export GOPATH="$TMP_DIR"',
+                '$TOOL get -d ' + get,
+            ]
+        if revision:
+            # Annoyingly -C does not work on git checkout :(
+            cmd.append('(cd %s && git checkout -q %s)' % (subdir, revision))
+        cmd.append('find . -name .git | xargs rm -rf')
+    if patch:
+        cmd.append('patch -s -d %s -p1 < ${TMP_DIR}/$SRCS' % subdir)
+    if strip:
+        cmd += ['rm -rf %s/%s' % (subdir, s) for s in strip]
+    # This is an awkward case, as noted above one may want to build a binary from an existing library
+    # so we have to special case here to allow them both to use the same source download.
+    get_rule = None if binary and deps else build_rule(
+        name = name,
+        tag = 'get',
+        srcs = [patch] if patch else [],
+        outs = [subdir],
+        tools = [CONFIG.GO_TOOL],
+        visibility = visibility,
+        building_description = 'Fetching...',
+        cmd = ' && '.join(cmd),
+        requires = ['go'],
+        test_only = test_only,
+        labels = labels,
+        hashes = hashes,
+        sandbox = False,
+    )
+
+    if install:
+        install = [i if i.startswith(get) else (get + '/' + i) for i in install]
+    else:
+        install = [get]
```

Comment:
should you be using `getroot` here instead of `get`?

Reply:
Not on line 606 for sure (it should install /... if specified). Maybe on 604; it doesn't really make sense to pass `install` with /... but `getroot` is probably better still.

Reply:
sorry, yes I meant mostly 604